### PR TITLE
Configurable OAuth2 content type for access token request 

### PR DESCRIPTION
**Pull request by pjagielski**

### CODE REVIEW

File: engine/security/src/test/scala/pl/touk/nussknacker/ui/security/http/RecordingSttpBackend.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.concurrent.atomic.AtomicReference
+
+import sttp.client.monad.MonadError
+import sttp.client.ws.WebSocketResponse
+import sttp.client.{Request, Response, SttpBackend}
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+
+class RecordingSttpBackend(delegate: SttpBackend[Future, Nothing, Nothing]) extends SttpBackend[Future, Nothing, Nothing] {
```
Comment (mproch) [inline engine/security/src/test/scala/pl/touk/nussknacker/ui/security/http/RecordingSttpBackend.scala:12]: Let's add comment that it comes from https://github.com/softwaremill/sttp/blob/master/core/src/main/scala/sttp/client3/testing/RecordingSttpBackend.scala and that it should be removed when we migrat to sttp3? Or if we added some methods - just a notice that it's based on sttp version

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala [hunk 2]
```diff
@@ -43,7 +51,7 @@ class OAuth2ClientApi[ProfileResponse: Decoder, AccessTokenResponse: Decoder]
     val headers = configuration.headers ++ Map(configuration.authorizationHeader -> s"Bearer $accessToken")
 
     basicRequest
-      .contentType(contentType.toString)
+      .contentType(MediaType.ApplicationJson)
```
Comment (mproch) [inline engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala:54]: this will always be json?

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala [hunk 3]
```diff
@@ -28,12 +28,20 @@ class OAuth2ClientApi[ProfileResponse: Decoder, AccessTokenResponse: Decoder]
...
+      basicRequest
+      .contentType(configuration.accessTokenRequestContentType)
       .response(asJson[AccessTokenResponse])
       .post(Uri(configuration.accessTokenUri))
       .headers(configuration.headers)
+
+    val contentType = MediaType.parse(configuration.accessTokenRequestContentType)
+    request = contentType match {
+      case Right(MediaType.ApplicationJson) => request.body(payload.asJson)
+      case _ => request.body(payload)
```
Comment (mproch) [inline engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala:41]: Will it work for all cases? Or just for x-www-form-urlencoded? Maybe we should explicitly enumerate types we can handle and throw exception otherwise?

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala [hunk 4]
```diff
@@ -26,7 +27,8 @@ case class OAuth2Configuration(method: AuthenticationMethod,
                                accessTokenParams: Map[String, String] = Map.empty,
                                authorizeParams: Map[String, String] = Map.empty,
                                headers: Map[String, String] = Map.empty,
-                               authorizationHeader: String = "Authorization"
+                               authorizationHeader: String = "Authorization",
+                               accessTokenRequestContentType: String = MediaType.ApplicationJson.toString()
```
Comment (mproch) [inline engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala:31]: please update Security.md (and ChangeLog.md)

**Review by mproch — COMMENTED** [review]
:+1: nice, I added minor comments

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala (hunk 2 above)
Comment (pjagielski) [reply engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala:54]: in fact, this should be "accept", since this is a GET request, but I think I'll just remove it..

Comment (pjagielski) [reply engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala:54]: hmm just got into this: https://docs.github.com/en/free-pro-team@latest/developers/apps/authorizing-oauth-apps#response

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (lciolecki) [reply engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala:31]: look at `profileFormat` it's enum.

File: engine/security/src/test/scala/pl/touk/nussknacker/ui/security/http/RecordingSttpBackend.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        case e: Exception =>
+          addInteraction(request, Failure(e))
+          responseMonad.error(e)
+      }) { response =>
+        addInteraction(request, Success(response))
+        response
+      }
+  }
+
+  override def openWebsocket[T, WS_RESULT](request: Request[T, Nothing], handler: Nothing): Future[WebSocketResponse[WS_RESULT]] = ???
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply engine/security/src/test/scala/pl/touk/nussknacker/ui/security/http/RecordingSttpBackend.scala:35]: Ok, right ;)

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (pjagielski) [reply engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala:31]: yeah, but `MediaType` is a case class from `sttp` we'll have to write custom (de)serializer... let's keep it in string

[the remark this answers is not in the document]
Comment (lciolecki) [reply engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala:31]: Uuu.. So be it.

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala (hunk 2 above)
Comment (pjagielski) [reply engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ClientApi.scala]: ok, but I see accept is set in build.sbt: https://github.com/TouK/nussknacker/blob/staging/build.sbt#L248

File: demo/docker/nk/conf/application.conf [hunk 6]
```diff
@@ -0,0 +1,133 @@
+storageDir: ./storage
```
Comment (mproch) [inline demo/docker/nk/conf/application.conf:1]: Why is this file here? Thought we'd stick with the defaults?

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/DefaultOAuth2ServiceFactory.scala [hunk 7]
```diff
@@ -33,19 +33,22 @@ class DefaultOAuth2Service[ProfileResponse: Decoder](clientApi: OAuth2ClientApi[
 
     configuration.jwt match {
       case None => profileRequestFuture
-      case Some(jwtConfiguration) => {
-        val profileFromJwtResult = new JwtValidator[ProfileResponse](jwtConfiguration).getProfileFromJwt(token)
+      case Some(jwtConfiguration) =>
```
Comment (mproch) [inline engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/DefaultOAuth2ServiceFactory.scala:36]: I thinks we can use 
case Some(jwtConfiguration) if jwtConfiguration.enabled => ...
case _ => profileRequestFuture

File: engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala [hunk 8]
```diff
@@ -38,9 +40,13 @@ case class OAuth2Configuration(method: AuthenticationMethod,
       .url)
   })
 
-  override def authSeverPublicKey: Option[PublicKey] = jwt.map(_.authServerPublicKey)
+  def jwtEnabled: Boolean = jwt.exists(_.enabled)
+
+  override def authSeverPublicKey: Option[PublicKey] = jwt.flatMap { jwt =>
+    if (!jwt.enabled) { None } else { Some(jwt.authServerPublicKey) }
```
Comment (mproch) [inline engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2Configuration.scala:46]: Can we use filter here? jwt.filter(_.enabled).map(_.authServerPublicKey)?

File: nussknacker-dist/src/universal/conf/dev-oauth2-application.conf [hunk 9]
```diff
@@ -0,0 +1,129 @@
+# This config is exposed in development dist and locally in development (se ui/server/runServer.sh)
```
Comment (mproch) [inline nussknacker-dist/src/universal/conf/dev-oauth2-application.conf:1]: Do we need yet another file here? Can we use dev-application and use env variables?

**Review by mproch — COMMENTED** [review]
Looks nice, I added some minor remarks

**Review by mproch — APPROVED** [review]
